File: crates/rullama-prompting/src/techniques.rs
//! Prompting Technique Definitions
//!
//! This module defines the 15 prompting techniques from the paper
//! "Adaptive Selection of Prompting Techniques" (arXiv:2510.18162),
//! with SEAL quality integration for intelligent technique filtering.

use serde::{Deserialize, Serialize};

/// Prompting technique categories from the paper
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum TechniqueCategory {
    /// Role assignment (e.g., "You are an expert...")
    RoleAssignment,
    /// Emotional stimulus (e.g., "This is important...")
    EmotionalStimulus,
    /// Reasoning techniques (e.g., Chain-of-Thought)
    Reasoning,
    /// Supporting techniques (e.g., Scratchpad)
    Others,
}

/// 15 prompting techniques from the paper (Table 1)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum PromptingTechnique {
    /// Role-playing persona assignment.
    RolePlaying,

    /// Emotion-based prompting stimulus.
    EmotionPrompting,
    /// Stress / urgency-based prompting stimulus.
    StressPrompting,

    /// Chain-of-thought reasoning.
    ChainOfThought,
    /// Logic-of-thought structured reasoning.
    LogicOfThought,
    /// Least-to-most decomposition.
    LeastToMost,
    /// Thread-of-thought sequential reasoning.
    ThreadOfThought,
    /// Plan-and-solve two-stage reasoning.
    PlanAndSolve,
    /// Skeleton-of-thought parallel generation.
    SkeletonOfThought,
    /// Scratchpad-based working memory.
    ScratchpadPrompting,

    /// Decomposed multi-step prompting.
    DecomposedPrompting,
    /// Ignore irrelevant conditions filtering.
    IgnoreIrrelevantConditions,
    /// Highlighted chain-of-thought.
    HighlightedCoT,
    /// Skills-in-context grounding.
    SkillsInContext,
    /// Automatic information filtering.
    AutomaticInformationFiltering,
}

impl PromptingTechnique {
    /// Parse technique from string ID
    pub fn parse_id(s: &str) -> Result<Self, &'static str> {
        match s.to_lowercase().as_str() {
            "role_playing" | "roleplaying" => Ok(Self::RolePlaying),
            "emotion_prompting" | "emotionprompting" => Ok(Self::EmotionPrompting),
            "stress_prompting" | "stressprompting" => Ok(Self::StressPrompting),
            "chain_of_thought" | "chainofthought" | "cot" => Ok(Self::ChainOfThought),
            "logic_of_thought" | "logicofthought" | "lot" => Ok(Self::LogicOfThought),
            "least_to_most" | "leasttomost" => Ok(Self::LeastToMost),
            "thread_of_thought" | "threadofthought" | "tot" => Ok(Self::ThreadOfThought),
            "plan_and_solve" | "planandsolve" => Ok(Self::PlanAndSolve),
            "skeleton_of_thought" | "skeletonofthought" | "sot" => Ok(Self::SkeletonOfThought),
            "scratchpad_prompting" | "scratchpadprompting" | "scratchpad" => {
                Ok(Self::ScratchpadPrompting)
            }
            "decomposed_prompting" | "decomposedprompting" => Ok(Self::DecomposedPrompting),
            "ignore_irrelevant_conditions" | "ignoreirrelevantconditions" => {
                Ok(Self::IgnoreIrrelevantConditions)
            }
            "highlighted_cot" | "highlightedcot" => Ok(Self::HighlightedCoT),
            "skills_in_context" | "skillsincontext" => Ok(Self::SkillsInContext),
            "automatic_information_filtering" | "automaticinformationfiltering" => {
                Ok(Self::AutomaticInformationFiltering)
            }
            _ => Err("Unknown technique"),
        }
    }

    /// Convert technique to string ID
    pub fn to_str(&self) -> &'static str {
        match self {
            Self::RolePlaying => "role_playing",
            Self::EmotionPrompting => "emotion_prompting",
            Self::StressPrompting => "stress_prompting",
            Self::ChainOfThought => "chain_of_thought",
            Self::LogicOfThought => "logic_of_thought",
            Self::LeastToMost => "least_to_most",
            Self::ThreadOfThought => "thread_of_thought",
            Self::PlanAndSolve => "plan_and_solve",
            Self::SkeletonOfThought => "skeleton_of_thought",
            Self::ScratchpadPrompting => "scratchpad_prompting",
            Self::DecomposedPrompting => "decomposed_prompting",
            Self::IgnoreIrrelevantConditions => "ignore_irrelevant_conditions",
            Self::HighlightedCoT => "highlighted_cot",
            Self::SkillsInContext => "skills_in_context",
            Self::AutomaticInformationFiltering => "automatic_information_filtering",
        }
    }
}

/// Complexity level for SEAL quality filtering
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum ComplexityLevel {
    /// Use when SEAL quality < 0.5 (simple, basic techniques)
    Simple,
    /// Use when SEAL quality 0.5-0.8 (moderate complexity)
    Moderate,
    /// Use when SEAL quality > 0.8 (advanced, sophisticated techniques)
    Advanced,
}

/// Task characteristics for technique matching
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum TaskCharacteristic {
    /// Requires multi-step reasoning chains.
    MultiStepReasoning,
    /// Involves numerical calculations.
    NumericalCalculation,
    /// Requires logical deduction.
    LogicalDeduction,
    /// Creative or open-ended generation.
    CreativeGeneration,
    /// Summarization of long context.
    LongContextSummarization,
    /// Spatial reasoning tasks.
    SpatialReasoning,
    /// Visual understanding tasks.
    VisualUnderstanding,
    /// Code generation tasks.
    CodeGeneration,
    /// Algorithmic problem solving.
    AlgorithmicProblem,
}

/// Metadata for each technique (SEAL-enhanced)
#[derive(Debug, Clone)]
pub struct TechniqueMetadata {
    /// The prompting technique this metadata describes.
    pub technique: PromptingTechnique,
    /// Category of the technique.
    pub category: TechniqueCategory,
    /// Human-readable name.
    pub name: String,
    /// Description of the technique.
    pub description: String,
    /// Template string for generating prompts.
    pub template: String,
    /// Task characteristics this technique works best for.
    pub best_for: Vec<TaskCharacteristic>,

    // SEAL integration
    /// Minimum SEAL quality to use this technique (0.0-1.0)
    pub min_seal_quality: f32,
    /// Complexity level for filtering
    pub complexity_level: ComplexityLevel,
    /// Can this technique be promoted to BKS?
    pub bks_promotion_eligible: bool,
}

impl TechniqueMetadata {
    /// Create new technique metadata
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        technique: PromptingTechnique,
        category: TechniqueCategory,
        name: impl Into<String>,
        description: impl Into<String>,
        template: impl Into<String>,
        best_for: Vec<TaskCharacteristic>,
        min_seal_quality: f32,
        complexity_level: ComplexityLevel,
        bks_promotion_eligible: bool,
    ) -> Self {
        Self {
            technique,
            category,
            name: name.into(),
            description: description.into(),
            template: template.into(),
            best_for,
            min_seal_quality,
            complexity_level,
            bks_promotion_eligible,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_technique_enum_all_variants() {
        // Ensure all 15 techniques can be created
        let techniques = vec![
            PromptingTechnique::RolePlaying,
            PromptingTechnique::EmotionPrompting,
            PromptingTechnique::StressPrompting,
            PromptingTechnique::ChainOfThought,
            PromptingTechnique::LogicOfThought,
            PromptingTechnique::LeastToMost,
            PromptingTechnique::ThreadOfThought,
            PromptingTechnique::PlanAndSolve,
            PromptingTechnique::SkeletonOfThought,
            PromptingTechnique::ScratchpadPrompting,
            PromptingTechnique::DecomposedPrompting,
            PromptingTechnique::IgnoreIrrelevantConditions,
            PromptingTechnique::HighlightedCoT,
            PromptingTechnique::SkillsInContext,
            PromptingTechnique::AutomaticInformationFiltering,
        ];
        assert_eq!(techniques.len(), 15);
    }

    #[test]
    fn test_technique_category_variants() {
        let categories = [
            TechniqueCategory::RoleAssignment,
            TechniqueCategory::EmotionalStimulus,
            TechniqueCategory::Reasoning,
            TechniqueCategory::Others,
        ];
        assert_eq!(categories.len(), 4);
    }

    #[test]
    fn test_complexity_level_ordering() {
        // Test complexity levels
        let simple = ComplexityLevel::Simple;
        let moderate = ComplexityLevel::Moderate;
        let advanced = ComplexityLevel::Advanced;

        // Ensure they can be compared
        assert_eq!(simple, ComplexityLevel::Simple);
        assert_eq!(moderate, ComplexityLevel::Moderate);
        assert_eq!(advanced, ComplexityLevel::Advanced);
    }

    #[test]
    fn test_technique_string_conversion() {
        // Test from_str and to_string
        let technique = PromptingTechnique::ChainOfThought;
        let serialized = serde_json::to_string(&technique).unwrap();
        let deserialized: PromptingTechnique = serde_json::from_str(&serialized).unwrap();
        assert_eq!(technique, deserialized);
    }

    #[test]
    fn test_technique_metadata_creation() {
        let metadata = TechniqueMetadata::new(
            PromptingTechnique::ChainOfThought,
            TechniqueCategory::Reasoning,
            "Chain-of-Thought",
            "Step-by-step reasoning",
            "Think step by step",
            vec![TaskCharacteristic::MultiStepReasoning],
            0.0,
            ComplexityLevel::Simple,
            true,
        );

        assert_eq!(metadata.technique, PromptingTechnique::ChainOfThought);
        assert_eq!(metadata.category, TechniqueCategory::Reasoning);
        assert_eq!(metadata.name, "Chain-of-Thought");
        assert_eq!(metadata.min_seal_quality, 0.0);
        assert_eq!(metadata.complexity_level, ComplexityLevel::Simple);
        assert!(metadata.bks_promotion_eligible);
    }

    #[test]
    fn test_task_characteristic_variants() {
        let characteristics = [
            TaskCharacteristic::MultiStepReasoning,
            TaskCharacteristic::NumericalCalculation,
            TaskCharacteristic::LogicalDeduction,
            TaskCharacteristic::CreativeGeneration,
            TaskCharacteristic::LongContextSummarization,
            TaskCharacteristic::SpatialReasoning,
            TaskCharacteristic::VisualUnderstanding,
        ];
        assert_eq!(characteristics.len(), 7);
    }
}

File: deno/packages/prompting/techniques.ts
/**
 * Definitions of the 15 prompting techniques from "Adaptive Selection of
 * Prompting Techniques" (arXiv:2510.18162). Provides the `PromptingTechnique`,
 * `TechniqueCategory`, `ComplexityLevel` and `TaskCharacteristic` string unions,
 * the `TECHNIQUE_METADATA` table describing each technique, and the lookup and
 * filter helpers (`getTechniqueMetadata`, `getTechniquesByCategory`,
 * `getTechniquesByComplexity`, `getTechniquesBySealQuality`, `parseTechniqueId`).
 * Equivalent to Rust's `rullama_prompting::techniques`.
 *
 * @module
 */

// ---------------------------------------------------------------------------
// Enums (represented as string unions for TypeScript ergonomics)
// ---------------------------------------------------------------------------

/** Prompting technique categories from the paper. */
export type TechniqueCategory =
  | "RoleAssignment"
  | "EmotionalStimulus"
  | "Reasoning"
  | "Others";

/** 15 prompting techniques from the paper (Table 1). */
export type PromptingTechnique =
  | "RolePlaying"
  | "EmotionPrompting"
  | "StressPrompting"
  | "ChainOfThought"
  | "LogicOfThought"
  | "LeastToMost"
  | "ThreadOfThought"
  | "PlanAndSolve"
  | "SkeletonOfThought"
  | "ScratchpadPrompting"
  | "DecomposedPrompting"
  | "IgnoreIrrelevantConditions"
  | "HighlightedCoT"
  | "SkillsInContext"
  | "AutomaticInformationFiltering";

/** Complexity level for SEAL quality filtering. */
export type ComplexityLevel =
  | "Simple"
  | "Moderate"
  | "Advanced";

/** Task characteristics for technique matching. */
export type TaskCharacteristic =
  | "MultiStepReasoning"
  | "NumericalCalculation"
  | "LogicalDeduction"
  | "CreativeGeneration"
  | "LongContextSummarization"
  | "SpatialReasoning"
  | "VisualUnderstanding"
  | "CodeGeneration"
  | "AlgorithmicProblem";

// ---------------------------------------------------------------------------
// All enum values (for iteration)
// ---------------------------------------------------------------------------

/** All 15 prompting techniques. */
export const ALL_TECHNIQUES: readonly PromptingTechnique[] = [
  "RolePlaying",
  "EmotionPrompting",
  "StressPrompting",
  "ChainOfThought",
  "LogicOfThought",
  "LeastToMost",
  "ThreadOfThought",
  "PlanAndSolve",
  "SkeletonOfThought",
  "ScratchpadPrompting",
  "DecomposedPrompting",
  "IgnoreIrrelevantConditions",
  "HighlightedCoT",
  "SkillsInContext",
  "AutomaticInformationFiltering",
] as const;

/** All technique categories. */
export const ALL_CATEGORIES: readonly TechniqueCategory[] = [
  "RoleAssignment",
  "EmotionalStimulus",
  "Reasoning",
  "Others",
] as const;

/** All complexity levels. */
export const ALL_COMPLEXITY_LEVELS: readonly ComplexityLevel[] = [
  "Simple",
  "Moderate",
  "Advanced",
] as const;

/** All task characteristics. */
export const ALL_TASK_CHARACTERISTICS: readonly TaskCharacteristic[] = [
  "MultiStepReasoning",
  "NumericalCalculation",
  "LogicalDeduction",
  "CreativeGeneration",
  "LongContextSummarization",
  "SpatialReasoning",
  "VisualUnderstanding",
  "CodeGeneration",
  "AlgorithmicProblem",
] as const;

// ---------------------------------------------------------------------------
// Technique ID string conversion (snake_case <-> enum)
// ---------------------------------------------------------------------------

const TECHNIQUE_TO_ID: Record<PromptingTechnique, string> = {
  RolePlaying: "role_playing",
  EmotionPrompting: "emotion_prompting",
  StressPrompting: "stress_prompting",
  ChainOfThought: "chain_of_thought",
  LogicOfThought: "logic_of_thought",
  LeastToMost: "least_to_most",
  ThreadOfThought: "thread_of_thought",
  PlanAndSolve: "plan_and_solve",
  SkeletonOfThought: "skeleton_of_thought",
  ScratchpadPrompting: "scratchpad_prompting",
  DecomposedPrompting: "decomposed_prompting",
  IgnoreIrrelevantConditions: "ignore_irrelevant_conditions",
  HighlightedCoT: "highlighted_cot",
  SkillsInContext: "skills_in_context",
  AutomaticInformationFiltering: "automatic_information_filtering",
};

const ID_TO_TECHNIQUE: Record<string, PromptingTechnique> = {};
for (const [technique, id] of Object.entries(TECHNIQUE_TO_ID)) {
  ID_TO_TECHNIQUE[id] = technique as PromptingTechnique;
  // Also map the camelCase (lowered) variant
  ID_TO_TECHNIQUE[technique.toLowerCase()] = technique as PromptingTechnique;
}
// Common abbreviations
ID_TO_TECHNIQUE["cot"] = "ChainOfThought";
ID_TO_TECHNIQUE["lot"] = "LogicOfThought";
ID_TO_TECHNIQUE["tot"] = "ThreadOfThought";
ID_TO_TECHNIQUE["sot"] = "SkeletonOfThought";
ID_TO_TECHNIQUE["scratchpad"] = "ScratchpadPrompting";

/** Convert a PromptingTechnique to its snake_case string ID. */
export function techniqueToId(technique: PromptingTechnique): string {
  return TECHNIQUE_TO_ID[technique];
}

/** Parse a string ID into a PromptingTechnique, or return undefined. */
export function parseTechniqueId(
  s: string,
): PromptingTechnique | undefined {
  return ID_TO_TECHNIQUE[s.toLowerCase()];
}

// ---------------------------------------------------------------------------
// Technique metadata
// ---------------------------------------------------------------------------

/** Metadata for each technique (SEAL-enhanced). */
export interface TechniqueMetadata {
  /** The prompting technique this metadata describes. */
  readonly technique: PromptingTechnique;
  /** Category of the technique. */
  readonly category: TechniqueCategory;
  /** Human-readable name. */
  readonly name: string;
  /** Description of the technique. */
  readonly description: string;
  /** Template string for generating prompts. */
  readonly template: string;
  /** Task characteristics this technique works best for. */
  readonly bestFor: readonly TaskCharacteristic[];
  /** Minimum SEAL quality to use this technique (0.0-1.0). */
  readonly minSealQuality: number;
  /** Complexity level for filtering. */
  readonly complexityLevel: ComplexityLevel;
  /** Can this technique be promoted to BKS? */
  readonly bksPromotionEligible: boolean;
}

// ---------------------------------------------------------------------------
// All 15 technique definitions (from TechniqueLibrary::new in Rust)
// ---------------------------------------------------------------------------

/** Complete registry of all 15 technique metadata definitions. */
export const TECHNIQUE_METADATA: ReadonlyMap<
  PromptingTechnique,
  TechniqueMetadata
> = new Map<PromptingTechnique, TechniqueMetadata>([
  // === Role Assignment (1 technique) ===
  [
    "RolePlaying",
    {
      technique: "RolePlaying",
      category: "RoleAssignment",
      name: "Role Playing",
      description: "Assign expert role to elicit domain-specific knowledge",
      template: "You are a {role} with expertise in {domain}. ",
      bestFor: ["MultiStepReasoning", "CodeGeneration", "AlgorithmicProblem"],
      minSealQuality: 0.0,
      complexityLevel: "Simple",
      bksPromotionEligible: true,
    },
  ],
  // === Emotional Stimulus (2 techniques) ===
  [
    "EmotionPrompting",
    {
      technique: "EmotionPrompting",
      category: "EmotionalStimulus",
      name: "Emotion Prompting",
      description: "Add emotional cues to increase engagement",
      template: "This is an important {task_type} that requires {quality}. ",
      bestFor: ["MultiStepReasoning", "NumericalCalculation"],
      minSealQuality: 0.0,
      complexityLevel: "Simple",
      bksPromotionEligible: true,
    },
  ],
  [
    "StressPrompting",
    {
      technique: "StressPrompting",
      category: "EmotionalStimulus",
      name: "Stress Prompting",
      description: "Induce moderate stress conditions for focus",
      template:
        "This task requires immediate attention and precision. Time is limited. ",
      bestFor: ["LogicalDeduction", "AlgorithmicProblem"],
      minSealQuality: 0.0,
      complexityLevel: "Simple",
      bksPromotionEligible: true,
    },
  ],
  // === Reasoning (7 techniques) ===
  [
    "ChainOfThought",
    {
      technique: "ChainOfThought",
      category: "Reasoning",
      name: "Chain-of-Thought",
      description: "Require explicit step-by-step reasoning",
      template: "Think step by step. Show your reasoning process clearly. ",
      bestFor: [
        "MultiStepReasoning",
        "LogicalDeduction",
        "NumericalCalculation",
      ],
      minSealQuality: 0.0,
      complexityLevel: "Simple",
      bksPromotionEligible: true,
    },
  ],
  [
    "LogicOfThought",
    {
      technique: "LogicOfThought",
      category: "Reasoning",
      name: "Logic-of-Thought",
      description: "Embed propositional logic for formal reasoning",
      template:
        "Use propositional logic notation. Let P, Q, R represent propositions. Apply logical inference rules. ",
      bestFor: ["LogicalDeduction", "MultiStepReasoning"],
      minSealQuality: 0.8,
      complexityLevel: "Advanced",
      bksPromotionEligible: true,
    },
  ],
  [
    "LeastToMost",
    {
      technique: "LeastToMost",
      category: "Reasoning",
      name: "Least-to-Most",
      description: "Decompose into simpler sub-problems progressively",
      template:
        "Break this problem into simpler sub-problems. Solve from simplest to most complex. ",
      bestFor: ["MultiStepReasoning", "AlgorithmicProblem"],
      minSealQuality: 0.5,
      complexityLevel: "Moderate",
      bksPromotionEligible: true,
    },
  ],
  [
    "ThreadOfThought",
    {
      technique: "ThreadOfThought",
      category: "Reasoning",
      name: "Thread-of-Thought",
      description: "Summarize long contexts progressively",
      template:
        "Summarize the context progressively as you reason through it. Maintain a running summary. ",
      bestFor: ["LongContextSummarization", "MultiStepReasoning"],
      minSealQuality: 0.7,
      complexityLevel: "Advanced",
      bksPromotionEligible: true,
    },
  ],
  [
    "PlanAndSolve",
    {
      technique: "PlanAndSolve",
      category: "Reasoning",
      name: "Plan-and-Solve",
      description: "Generate execution plan first, then solve step by step",
      template:
        "First, devise a plan. Then, solve the problem step by step according to the plan. ",
      bestFor: [
        "MultiStepReasoning",
        "LogicalDeduction",
        "AlgorithmicProblem",
      ],
      minSealQuality: 0.5,
      complexityLevel: "Moderate",
      bksPromotionEligible: true,
    },
  ],
  [
    "SkeletonOfThought",
    {
      technique: "SkeletonOfThought",
      category: "Reasoning",
      name: "Skeleton-of-Thought",
      description: "Generate skeleton, then fill details",
      template:
        "First, generate a skeleton outline. Then, fill in the details for each part. ",
      bestFor: ["CreativeGeneration", "CodeGeneration"],
      minSealQuality: 0.7,
      complexityLevel: "Advanced",
      bksPromotionEligible: true,
    },
  ],
  [
    "ScratchpadPrompting",
    {
      technique: "ScratchpadPrompting",
      category: "Reasoning",
      name: "Scratchpad Prompting",
      description: "Provide draft space for intermediate steps",
      template:
        "Use the following scratchpad format for intermediate calculations:\n<scratchpad>\n[Your work here]\n</scratchpad>\n",
      bestFor: ["NumericalCalculation", "AlgorithmicProblem"],
      minSealQuality: 0.0,
      complexityLevel: "Simple",
      bksPromotionEligible: true,
    },
  ],
  // === Others (5 techniques) ===
  [
    "DecomposedPrompting",
    {
      technique: "DecomposedPrompting",
      category: "Others",
      name: "Decomposed Prompting",
      description: "Break into sub-tasks explicitly",
      template:
        "Decompose this task into independent sub-tasks. Solve each sub-task separately. ",
      bestFor: ["MultiStepReasoning", "AlgorithmicProblem"],
      minSealQuality: 0.5,
      complexityLevel: "Moderate",
      bksPromotionEligible: true,
    },
  ],
  [
    "IgnoreIrrelevantConditions",
    {
      technique: "IgnoreIrrelevantConditions",
      category: "Others",
      name: "Ignore Irrelevant Conditions",
      description: "Detect and disregard noise in the problem",
      template:
        "Identify and ignore any irrelevant information. Focus only on what's essential. ",
      bestFor: ["LogicalDeduction", "MultiStepReasoning"],
      minSealQuality: 0.6,
      complexityLevel: "Moderate",
      bksPromotionEligible: true,
    },
  ],
  [
    "HighlightedCoT",
    {
      technique: "HighlightedCoT",
      category: "Others",
      name: "Highlighted CoT",
      description: "Highlight essential information before reasoning",
      template:
        "First, highlight the essential information. Then, reason step by step based on the highlights. ",
      bestFor: ["MultiStepReasoning", "LogicalDeduction"],
      minSealQuality: 0.5,
      complexityLevel: "Moderate",
      bksPromotionEligible: true,
    },
  ],
  [
    "SkillsInContext",
    {
      technique: "SkillsInContext",
      category: "Others",
      name: "Skills-in-Context",
      description: "Compose basic skills for complex tasks",
      template:
        "Identify the basic skills required. Compose them systematically to solve the task. ",
      bestFor: ["AlgorithmicProblem", "CodeGeneration"],
      minSealQuality: 0.7,
      complexityLevel: "Advanced",
      bksPromotionEligible: true,
    },
  ],
  [
    "AutomaticInformationFiltering",
    {
      technique: "AutomaticInformationFiltering",
      category: "Others",
      name: "Automatic Information Filtering",
      description: "Preprocess to remove irrelevant information",
      template:
        "Filter the input to retain only relevant information before processing. ",
      bestFor: ["LongContextSummarization", "LogicalDeduction"],
      minSealQuality: 0.6,
      complexityLevel: "Moderate",
      bksPromotionEligible: true,
    },
  ],
]);

// ---------------------------------------------------------------------------
// Convenience accessors
// ---------------------------------------------------------------------------

/** Get metadata for a specific technique. */
export function getTechniqueMetadata(
  technique: PromptingTechnique,
): TechniqueMetadata | undefined {
  return TECHNIQUE_METADATA.get(technique);
}

/** Get all technique metadata entries. */
export function getAllTechniqueMetadata(): TechniqueMetadata[] {
  return [...TECHNIQUE_METADATA.values()];
}

/** Get techniques filtered by minimum SEAL quality score. */
export function getTechniquesBySealQuality(
  sealQuality: number,
): TechniqueMetadata[] {
  return getAllTechniqueMetadata().filter(
    (t) => t.minSealQuality <= sealQuality,
  );
}

/** Get techniques by category. */
export function getTechniquesByCategory(
  category: TechniqueCategory,
): TechniqueMetadata[] {
  return getAllTechniqueMetadata().filter((t) => t.category === category);
}

/** Get techniques by complexity level. */
export function getTechniquesByComplexity(
  level: ComplexityLevel,
): TechniqueMetadata[] {
  return getAllTechniqueMetadata().filter(
    (t) => t.complexityLevel === level,
  );
}

/** Count techniques by complexity level. */
export function countByComplexity(level: ComplexityLevel): number {
  return getTechniquesByComplexity(level).length;
}
